### PR TITLE
CuTe-DSL modular prefill: sliding-window support + band-mask performance

### PR DESCRIPTION
**Pull request by pgera**
# CuTe-DSL modular prefill: sliding-window support + band-mask performance

## Summary

The modular CuTe-DSL prefill (SM100) mapped `causal=True + window_left >= 0`
to a plain causal mask — wrong results, and full O(s²/2) work. This PR fixes
the mask semantics with composable per-row bands, removes the banded regime's
dead work, makes window sizes runtime arguments (one compiled kernel per mask
kind), and routes windowed plans to this path. It also adds native LSE
output and mixed K/V dtype support (bf16/fp16 Q/K + fp8 V) to the modular
kernel, so windowed plans never leave it at run time.

## Changes

- **Composable band masks (correctness fix)**: the `MaskType` enum becomes a
  band `[lo(q), hi(q))` — causal composes with `window_left`; `window_left`
  alone is a left-bounded non-causal window (FA2 semantics); new
  `window_right` plan() parameter for symmetric windows. Loader / MMA /
  softmax / correction all derive trip counts from the same band helpers.
- **Dead work removal**: per-half band iteration (head/tail peel) — each
  128-row half of the CTA runs exactly its own band's trips instead of the
  union with ~1 dead step per stage; warp-ballot rescale skip in correction
  (skips the O-tile TMEM roundtrip when the factor is exactly 1).
- **Mask codegen**: hoisted row-invariant bounds, trace-time-constant K
  coordinates, `cutlass.select_` for masked writes; one shared path for all
  four mask kinds (avoids an MLIR lowering pathology that caused register
  spills on the release toolchain).
- **Runtime window values**: `MaskSpec` is two presence bits; window sizes
  are runtime `Int32` kernel args, causal is a right bound with value 0, and
  the `k < seqlen_k` tail check is always on. Mask compile matrix: 4 kernels.
- **Scheduling by mask kind**: banded masks default to non-persistent launch,
  unmasked stays persistent (`FLASHINFER_CUTE_PREFILL_PERSISTENT` overrides).
- **`sm_scale` default fix**: cute-dsl wrapper plan() now defaults to
  `1/sqrt(head_dim_qk)` (was 1.0, silently unscaled).
- **Vendored FMHA windowed compile fix**: the #3857 path crashed on any
  `window_left >= 0` (no window axis in the trace signature, compile cache,
  or artifact names); it now JIT-compiles with window-presence bits.
- **Native LSE**: `run(return_lse=True)` is computed in the correction warp
  from the final (row_sum, row_max) — log2-domain, matching the vendored
  kernel's convention. Zero cost when off (the non-LSE kernel's SASS is
  byte-identical) and ≲1% when on. Supported for score-mod and sink
  variants; logits-transform variants (no softmax) raise.
- **Mixed K/V dtypes**: bf16/fp16 Q/K with fp8 V. P converts to V's dtype
  for the PV MMA, V slots on the shared K/V smem ring re-arm their barrier
  with V's byte count, and the wrapper JIT-compiles per-V-dtype kernel
  variants at run time (V's dtype is a run-time property; plan()'s
  `kv_data_type` declares K and V together).
- **Routing**: `variant is None and head_dim_qk == 128 and window_left < 0`
  routes to the #3857 vendored kernel; everything else — windows, variants,
  other head dims, any V dtype, `return_lse` — runs the modular path. The
  gate is decided entirely at plan(); there is no run-time fallback.

## Correctness

New causal+window, left-only, and symmetric-window tests in
`tests/attention/test_modular_fmha_prefill.py`, including windowed masks
over mixed-length ragged batches (each sequence gets its own band
geometry; short sequences exercise the item-skip path). Validated on B200 against an
exact f32 reference over {8/8, 32/8, 64/8, 32/32} heads × 5 shapes × 4 masks
in both scheduler modes (152/152), with bitwise-identical outputs across
scheduler modes.

Also new: uniform-fp8 (e4m3) tests calibrated against the vendored kernel's
fp8 error on identical inputs (both ~0.066 max vs f32); mixed bf16-QK/fp8-V
tests against a dequantized-V f32 reference (~0.05 max — S stays bf16-exact);
LSE tests (log2-domain, f32 logsumexp references) covering windowed, ragged
varlen, ALiBi, and sink variants, with the output tensor bit-identical
between LSE and non-LSE runs.

## Performance vs trtllm-gen

B200 (SM100a), torch 2.11+cu130, `nvidia-cutlass-dsl` 4.6.0. CUPTI
kernel-only, median of 30 iters, cold L2; each cell benches both backends
back-to-back in one process, median of 3 passes, single node and session.
GQA 32/8 heads, head_dim 128, bf16. cute-dsl =
`BatchPrefillWithRaggedKVCacheWrapper(backend="cute-dsl")`, i.e. what the
public wrapper runs per mask kind: vendored route for unmasked/causal,
modular kernel for windows. trtllm-gen = paged wrapper, HND, page 16,
preallocated output. Ratio = cute-dsl / trtllm (< 1: cute-dsl faster).
GPU clocks are not lockable on these nodes, so absolute times are only
comparable within a table (each table is built from back-to-back
same-process measurements; the ratios are clock-robust).

Modular vs vendored, why windows route to the modular kernel (B=1 S=16K,
median of 3 passes):

| config | modular (µs) | vendored (µs) | modular/vendored |
|--|--:|--:|:--:|
| unmasked | 3540.4 | 2893.8 | 1.22 |
| causal | 1808.3 | 1532.7 | 1.18 |
| causal + w=127 | 265.0 | 359.8 | 0.74 |
| causal + w=511 | 314.4 | 620.9 | 0.51 |
| causal + w=1023 | 412.0 | 748.1 | 0.55 |

Shape grid (increasing S; per shape: unmasked, causal, windows in increasing
size). The kernel column shows which kernel the wrapper routes to: windowed
rows run this PR's modular path; unmasked/causal rows run the #3857 vendored
kernel and are included for context only.

| B | S | mask | kernel | cute-dsl (µs) | trtllm (µs) | ratio |
|--:|--:|--|--|--:|--:|:--:|
| 64 | 512 | unmasked | vendored | 330.9 | 280.8 | 1.18 |
| 64 | 512 | causal | vendored | 511.7 | 269.4 | 1.90 |
| 64 | 512 | w=256 | modular | 570.9 | 300.6 | 1.90 |
| 16 | 1024 | unmasked | vendored | 249.9 | 236.4 | 1.06 |
| 16 | 1024 | causal | vendored | 318.9 | 185.7 | 1.72 |
| 16 | 1024 | w=256 | modular | 302.7 | 169.2 | 1.79 |
| 1 | 4096 | unmasked | vendored | 201.8 | 227.4 | 0.89 |
| 1 | 4096 | causal | vendored | 159.3 | 150.0 | 1.06 |
| 1 | 4096 | w=256 | modular | 76.4 | 52.5 | 1.46 |
| 1 | 4096 | w=1024 | modular | 115.6 | 87.9 | 1.32 |
| 1 | 4096 | w=2048 | modular | 151.1 | 121.0 | 1.25 |
| 4 | 4096 | unmasked | vendored | 824.4 | 825.0 | 0.99 |
| 4 | 4096 | causal | vendored | 557.9 | 483.7 | 1.15 |
| 4 | 4096 | w=256 | modular | 288.3 | 172.4 | 1.67 |
| 4 | 4096 | w=1024 | modular | 428.5 | 284.5 | 1.51 |
| 4 | 4096 | w=2048 | modular | 519.2 | 399.7 | 1.30 |
| 1 | 16384 | unmasked | vendored | 2845.0 | 3376.5 | 0.86 |
| 1 | 16384 | causal | vendored | 1568.4 | 1758.7 | 0.86 |
| 1 | 16384 | w=256 | modular | 292.8 | 178.1 | 1.62 |
| 1 | 16384 | w=1024 | modular | 415.0 | 304.8 | 1.36 |
| 1 | 16384 | w=2048 | modular | 576.9 | 480.3 | 1.20 |
| 1 | 16384 | w=4096 | modular | 881.7 | 808.0 | 1.09 |
| 1 | 16384 | w=8192 | modular | 1402.0 | 1314.6 | 1.06 |
| 1 | 32768 | unmasked | vendored | 12151.0 | 13847.2 | 0.88 |
| 1 | 32768 | causal | vendored | 6353.5 | 7176.0 | 0.89 |
| 1 | 32768 | w=256 | modular | 580.1 | 334.8 | 1.77 |
| 1 | 32768 | w=1024 | modular | 798.7 | 603.6 | 1.33 |
| 1 | 32768 | w=2048 | modular | 1145.3 | 983.6 | 1.16 |
| 1 | 32768 | w=4096 | modular | 1955.4 | 1740.8 | 1.11 |
| 1 | 32768 | w=8192 | modular | 3443.3 | 3392.4 | 1.01 |

Window-size sweep (B=1, S=16384):

| window | cute-dsl (µs) | trtllm (µs) | ratio |
|--:|--:|--:|:--:|
| 127 | 236.2 | 145.6 | 1.62 |
| 255 | 269.7 | 167.9 | 1.61 |
| 511 | 312.9 | 216.8 | 1.44 |
| 1023 | 411.3 | 304.6 | 1.35 |
| 2047 | 578.0 | 480.0 | 1.20 |
| 4095 | 883.8 | 804.9 | 1.10 |
| 8191 | 1387.7 | 1320.7 | 1.04 |

Mixed V dtype (bf16 Q/K + fp8 V, B=1 S=8192, same protocol): mixed inputs
force the vendored path onto JIT compilation (its prebuilt artifact matrix
has no mixed-dtype axis), where the modular kernel also wins causal:

| config | modular (µs) | vendored-JIT (µs) | modular/vendored |
|--|--:|--:|:--:|
| unmasked | 739.4 | 706.3 | 1.05 |
| causal | 481.8 | 548.1 | 0.88 |
| causal + w=127 | 117.7 | 183.9 | 0.64 |
| causal + w=1023 | 208.4 | 395.2 | 0.53 |

## Summary by CodeRabbit

* **New Features**
  * Added flexible attention band masking, including causal, symmetric, left-only, and right-bounded sliding windows.
  * Added optional log-sum-exp (LSE) output, including variable-length, ALiBi, and attention-sink scenarios.
  * Added support for mixed Q/K/V data types in prefill attention.
  * Added automatic scale selection when no scale is provided.
* **Bug Fixes**
  * Improved kernel selection and fallback behavior for windowed attention and unsupported data-type combinations.
* **Tests**
  * Expanded coverage for masking, FP8, mixed dtypes, LSE results, and ragged inputs.

### CODE REVIEW

File: flashinfer/attention/cute_dsl/fmha.py [hunk 1]
```diff
@@ -403,6 +403,13 @@ def cute_dsl_fmha_ragged_prefill(
         try:
             if q.dtype != k.dtype or k.dtype != v.dtype:
                 raise RuntimeError("Separate dtype_qk / dtype_vo requires JIT")
```
[the remark this answers is not in the document]
Comment (pgera) [reply flashinfer/attention/cute_dsl/fmha.py]: Fixed in f13d017b: `q.dtype != k.dtype` now raises a hard `ValueError` before artifact/JIT resolution (the JIT compile builds both Q and K fake tensors from a single `qk_dtype`, so it could never serve that case); only a separate V dtype falls back to JIT.

**Comment by pgera** [comment]
@flashinfer-bot run

File: flashinfer/cute_dsl/attention/fusion/mask.py [hunk 2]
```diff
@@ -17,6 +34,324 @@
...
+
+@cute.jit
+def apply_mask(
+    spec: MaskSpec,
+    acc_qk: cute.Tensor,
+    index_qk: cute.Tensor,
+    seqlen_k: Int32,
+    causal_offset: Int32,
+    index_qk_static: cute.Tensor,
+    window_left: Int32,
```
Comment (PetersonGuo) [inline flashinfer/cute_dsl/attention/fusion/mask.py:290]: window_left and window_right would be layer constant. is there a reason why it was left as a runtime parameter instead? Could we compile kernels with the layer shapes instead?

File: flashinfer/cute_dsl/attention/wrappers/batch_prefill.py [hunk 3]
```diff
@@ -206,10 +207,11 @@ def plan(
         head_dim_qk,
         head_dim_vo=None,
         causal=True,
-        sm_scale=1.0,
+        sm_scale=None,
         q_data_type=torch.float16,
         kv_data_type=torch.float16,
         window_left: int = -1,
+        window_right: int = -1,
```
Comment (PetersonGuo) [inline flashinfer/cute_dsl/attention/wrappers/batch_prefill.py]: most models are still causal and window_right=0 represents this. Can we default this to 0 instead and remove the causal bool?

File: tests/attention/test_modular_fmha_prefill.py [hunk 4]
```diff
@@ -829,7 +835,13 @@ def attention_sliding_window_ref(
     qk_offset = kv_len - qo_len
     q_idx = torch.arange(qo_len, device=q.device).unsqueeze(1)
     k_idx = torch.arange(kv_len, device=q.device).unsqueeze(0)
-    mask = torch.abs(k_idx - (q_idx + qk_offset)) <= window_left
+    mask = torch.ones(qo_len, kv_len, dtype=torch.bool, device=q.device)
+    if causal:
+        mask &= k_idx <= q_idx + qk_offset
+    elif window_right >= 0:
```
Comment (PetersonGuo) [inline tests/attention/test_modular_fmha_prefill.py]: The causal branch looks redundant here

File: flashinfer/cute_dsl/attention/fusion/mask.py (hunk 2 above)
Comment (PetersonGuo) [reply flashinfer/cute_dsl/attention/fusion/mask.py:290]: would also remove the need for MaskSpec

**Review by PetersonGuo — CHANGES_REQUESTED** [review]

File: flashinfer/prefill.py [hunk 5]
```diff
@@ -3502,7 +3502,14 @@ def plan(
...
             )
             # Route compatible calls to faster trtllm CuTe DSL FMHA kernels.
-            self._cute_dsl_use_fmha = variant is None and head_dim_qk == 128
+            # Sliding-window plans stay on the modular cute-dsl path: it
+            # measures faster than the FMHA route on windows (its banded
+            # masks skip dead work), and the FMHA route's prebuilt artifact
+            # matrix has no windowed variants (windowed calls there need a
+            # JIT compile).
+            self._cute_dsl_use_fmha = (
+                variant is None and head_dim_qk == 128 and window_left < 0
```
Comment (saltyminty) [inline flashinfer/prefill.py:3511]: The modular cutedsl path isn't fully featured, i.e. it doesn't support return_lse and requires QKV to have the same dtype, so the new `window_left < 0` could lead to a unsupported regression compared to the previous path. I believe this warrants additional guards before choosing `self._cute_dsl_use_fmha`

File: flashinfer/cute_dsl/attention/wrappers/batch_prefill.py [hunk 6]
```diff
@@ -230,14 +232,21 @@ def plan(
...
         q_data_type : torch.dtype
             Data type for queries (float16, bfloat16, or float8_e4m3fn).
         kv_data_type : torch.dtype
             Data type for keys/values.
         window_left : int
-            Sliding window size. -1 disables sliding window.
+            Max lookback distance: query ``q`` attends to keys
+            ``k >= q + (kv_len - qo_len) - window_left``. -1 = unbounded.
+            Composes with ``causal`` (causal sliding window attention).
+        window_right : int
```
Comment (saltyminty) [inline flashinfer/cute_dsl/attention/wrappers/batch_prefill.py:266]: For backward compatibility, we should be appending all arguments to the end until the next major release.

**Comment by saltyminty** [comment]
/bot run tests/attention/test_modular_fmha_prefill.py tests/attention/test_cute_dsl_fmha_backend.py

File: flashinfer/prefill.py (hunk 5 above)
Comment (pgera) [reply flashinfer/prefill.py:3511]: I added LSE support and mixed KV types ( V= fp8, K = bf16) to the modular path. The features should be at parity now.

File: flashinfer/cute_dsl/attention/wrappers/batch_prefill.py (hunk 6 above)
Comment (pgera) [reply flashinfer/cute_dsl/attention/wrappers/batch_prefill.py:266]: Ack

File: flashinfer/cute_dsl/attention/wrappers/batch_prefill.py (hunk 3 above)
Comment (pgera) [reply flashinfer/cute_dsl/attention/wrappers/batch_prefill.py]: The causal bool is a part of the plan API shared across diff backends. So didn't want to break that.

File: flashinfer/cute_dsl/attention/fusion/mask.py (hunk 2 above)
Comment (pgera) [reply flashinfer/cute_dsl/attention/fusion/mask.py:290]: The window bounds can be compiled in, but it doesn't show any perf benefits. The core logic depends on the existence of windows. Keeping them runtime vals means that we don't need to JIT compile a new kernel for a different bound. There are 2 flags, so the universe is just 4 types of kernels.

**Comment by pgera** [comment]
@flashinfer-bot run

File: tests/attention/test_modular_fmha_prefill.py (hunk 4 above)
Comment (pgera) [reply tests/attention/test_modular_fmha_prefill.py]: Ack

**Review by xrq-phys — APPROVED** [review]
`fmha.py` runner changes LGTM. Thanks so much for spotting this!

**Review by saltyminty — APPROVED** [review]
